x-pack/plugins/fleet/server/services/agents/reassign.ts

```diff
@@ -113,16 +121,11 @@ export async function reassignAgents(
         throw new AgentReassignmentError(`${agent.id} is already assigned to ${newAgentPolicyId}`);
       }
 
-      const isAllowed = await reassignAgentIsAllowed(
-        soClient,
-        esClient,
-        agent.id,
-        newAgentPolicyId
-      );
-      if (isAllowed) {
-        return agent;
+      // by kuery already filtered out managed agents
+      if (!('kuery' in options)) {
+        await reassignAgentIsAllowed(soClient, esClient, agent.id, newAgentPolicyId);
```

Comment:
I am wondering if you have considered for that to change the signature of `reassignAgentIsAllowed` to pass a policy, we will have not to pass `withoutManaged` and an soClient to getAgents and it seems to me that it easier to test and more maintanable than a kql query.

1.  you will retrieve all the unique policy id from the agents list
2. you will fetch all the policies
3. you will pass the policy to `reassignAgentIsAllowed`

What do you think?

Reply:
Yes, I saw similar logic in bulk upgrade implementation.
I thought it's more efficient to do the managed filter on es query rather than in memory. I can try this and compare the performance, if there is no big difference, we can go with the in-memory filter version.

One disadvantage of filtering out managed in es query is that we can't return in response the managed agents which were left out.

Reply:
I did some testing locally and the in-memory filter seems much faster, I'll update the pr.
It looks like doing a terms query with multiple values on agents index is slower than I expected.

Reply:
Thanks for testing this, I think it will be easier to maintain too.